File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {
    const permutation = [];
    const res = [];
    const chosen = new Set();
    const search = function(){
        if(permutation.length === nums.length) res.push(Array.from(permutation));
        else{
            for(let i = 0; i<nums.length; i++){
                if(chosen.has(nums[i])) continue;
                chosen.add(nums[i]);
                permutation.push(nums[i]);
                search();
                chosen.delete(nums[i]);
                permutation.pop();
            }
        }
    }
    search();
    return res;
};

File: 0046-permutations/0046-permutations.cpp
class Solution {
private:
    vector<vector<int>> res;
    vector<int> permutation;
    unordered_set<int> chosen;
public:
    void search(vector<int>& nums){
        int n = nums.size();
        if(permutation.size() == n){
            // goal
            res.push_back(permutation);
        }else{
            for(int i = 0; i<n; i++){
                int choice = nums[i];
                if(chosen.find(choice) != chosen.end()) continue;
                permutation.push_back(choice);
                chosen.insert(choice);
                search(nums);
                chosen.erase(choice);
                permutation.pop_back();
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        search(nums);
        return res;
    }
};